libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async
+    static_assert(_Copy_size == 4 || _Copy_size == 8 || _Copy_size == 16, "cp.async.shared.global requires a copy size of 4, 8, or 16.");
+    asm volatile(
+        "cp.async.ca.shared.global [%0], [%1], %2, %2;"
+        :
+        : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__dest))),
+          "l"(static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__src))),
+          "n"(_Copy_size)
+        : "memory");
+}
+
+template <>
+inline __device__
+void __cp_async_shared_global<16>(char * __dest, const char * __src) {
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async
+    // When copying 16 bytes, it is possible to skip L1 cache (.cg).
+    asm volatile(
+        "cp.async.cg.shared.global [%0], [%1], %2, %2;"
+        :
+        : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__dest))),
+          "l"(static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__src))),
+          "n"(16)
+        : "memory");
+}
+
+template <size_t _Alignment, typename _Group>
+inline __device__
+void __cp_async_shared_global_impl(_Group __g, char * __dest, const char * __src, _CUDA_VSTD::size_t __size) {
+    static_assert(4 <= _Alignment, "cp.async requires at least 4-byte alignment");
+
+    // Maximal copy size is 16.
+    constexpr int __copy_size = (16 < _Alignment) ? 16 : _Alignment;
+    // We use an int offset here, because we are copying to shared memory,
+    // which is easily addressable using int.
+    const int __group_size = __g.size();
+    const int __group_rank = __g.thread_rank();
+    const int __stride = __group_size * __copy_size;
+    for (int __offset = __group_rank *  __copy_size; __offset < __size; __offset += __stride) {
+        __cp_async_shared_global<__copy_size>(__dest + __offset, __src + __offset);
     }
+}
+#endif // __CUDA_MINIMUM_ARCH__
+
+template <size_t _Copy_size>
+struct __copy_chunk {
+    _ALIGNAS(_Copy_size) char data[_Copy_size];
 };
 
-template<_CUDA_VSTD::size_t _Native_alignment, typename _Group, typename _Sync>
-_LIBCUDACXX_INLINE_VISIBILITY
-void inline __memcpy_async_sm_dispatch(
-        _Group const & __group, char * __destination, char const * __source,
-        _CUDA_VSTD::size_t __size, _Sync & __sync, async_contract_fulfillment & __is_async) {
-    // Broken out of __memcpy_async to avoid nesting dispatches
-    NV_DISPATCH_TARGET(
-        NV_PROVIDES_SM_80,
-            __is_async = __memcpy_async_impl<16>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size());
-    )
+template <size_t _Alignment, typename _Group>
+inline __host__ __device__
+void __cp_async_fallback(_Group __g, char * __dest, const char * __src, _CUDA_VSTD::size_t __size) {
+    // constexpr _CUDA_VSTD::size_t __copy_size = (16 < _Alignment) ? 16 : _Alignment;
```

Comment:
Is that a remnant from a previous implementation?
```suggestion
```

Reply:
No, this is something that was left over from debugging. Good catch! I fixed it now.